In [1]:
import pandas

In [2]:
base_path_a = "Datasets - Teste Data Engineer - Passei Direto/BASE A/{0}.json"
courses_DF = pandas.read_json(base_path_a.format("courses"))
sessions_DF = pandas.read_json(base_path_a.format("sessions"))
student_follow_subject_DF = pandas.read_json(base_path_a.format("student_follow_subject"))
students_DF = pandas.read_json(base_path_a.format("students"))
subscriptions_DF = pandas.read_json(base_path_a.format("subscriptions"))
universities_DF = pandas.read_json(base_path_a.format("universities"))


In [84]:
subjects_DF = pandas.read_json(base_path_a.format("subjects"))

# Data Exploration

In [3]:
student_follow_subject_DF.head(10)

FollowDate  \
0  2015-09-07 15:49:12.000000   
1  2015-09-07 15:49:12.000000   
2  2015-09-07 15:49:12.373000   
3  2015-09-07 15:49:12.000000   
4  2015-09-07 15:49:12.000000   
5  2015-12-07 16:17:54.000000   
6  2015-12-07 16:17:54.000000   
7  2015-12-07 16:17:54.000000   
8  2015-12-07 16:17:54.000000   
9  2015-12-07 16:17:54.000000   

                                           StudentId  SubjectId  
0  0cade9bf00234e3789afc76250e4042bc1b70784ae7004...     682889  
1  0cade9bf00234e3789afc76250e4042bc1b70784ae7004...     684377  
2  0cade9bf00234e3789afc76250e4042bc1b70784ae7004...     698583  
3  0cade9bf00234e3789afc76250e4042bc1b70784ae7004...     757882  
4  0cade9bf00234e3789afc76250e4042bc1b70784ae7004...    1910475  
5  8a501cab6c0a5a7e98a86f51a6264d81cdb562f4e99106...     673956  
6  8a501cab6c0a5a7e98a86f51a6264d81cdb562f4e99106...     674245  
7  8a501cab6c0a5a7e98a86f51a6264d81cdb562f4e99106...     681478  
8  8a501cab6c0a5a7e98a86f51a6264d81cdb562f4e99106...     681925  
9  8a501cab6c0a5a7e98a86f51a6264d81cdb562f4e99106...     693487

In [3]:
students_DF.dtypes

City              object
CourseId           int64
Id                object
RegisteredDate    object
SignupSource      object
State             object
StudentClient     object
UniversityId       int64
dtype: object

In [4]:
students_DF.head(3)

City  CourseId  \
0  Rio de Janeiro   1199555   
1  Rio de Janeiro   1199521   
2             NaN   1199517   

                                                  Id  \
0  0cade9bf00234e3789afc76250e4042bc1b70784ae7004...   
1  8a501cab6c0a5a7e98a86f51a6264d81cdb562f4e99106...   
2  b8a39150d98d746853087a4107e9290976cb6c99d16a29...   

               RegisteredDate SignupSource           State StudentClient  \
0  2012-09-06 17:27:11.370000     Facebook  Rio de Janeiro           NaN   
1  2012-09-05 15:31:08.090000     Facebook  Rio de Janeiro           NaN   
2  2012-08-17 10:00:04.270000     Facebook             NaN           NaN   

   UniversityId  
0        664704  
1        664704  
2        661625

In [28]:
students_DF.State.value_counts()

Rio de Janeiro         5168
São Paulo              3870
Minas Gerais           2417
Paraná                 1438
Bahia                  1354
Ceará                  1236
Rio Grande do Sul      1192
Pernambuco              954
Santa Catarina          876
Goiás                   815
Espírito Santo          595
Pará                    526
Maranhão                507
Sergipe                 427
Mato Grosso do Sul      421
Distrito Federal        385
Rio Grande do Norte     367
Paraíba                 345
Amazonas                342
Alagoas                 331
Piauí                   282
Mato Grosso             265
Tocantins               138
Amapá                   136
Rondônia                108
Roraima                  59
Acre                     53
Name: State, dtype: int64

In [23]:
students_DF["Id"].nunique()

60000

In [18]:
students_DF.shape

(60000, 8)

Pela descrição dos dados, existem alguns aqui que podem ser melhor utilizados em outros tipos, como "RegisteredDate" pode ser lido como timestamp e "City", "SignupSource"como categórico.

In [30]:
courses_DF.head(10)

Id                            Name
0  1199555             Engenharia Elétrica
1  1199521  Economia / Ciências Econômicas
2  1199517                         Direito
3  1199491             Ciências Ambientais
4  1199573              Engenharia Química
5  1199553          Engenharia de Produção
6  1199536            Engenharia Ambiental
7  1199725                         Química
8  1199453                   Administração
9  1199701                     Odontologia

In [31]:
courses_DF.dtypes

Id       int64
Name    object
dtype: object

In [32]:
courses_DF.shape

(471, 2)

Posso modificar o Name para categórico, assim dá para fazer algumas métricas interessantes utilizando pandas.

In [8]:
sessions_DF.dtypes

SessionStartTime    object
StudentClient       object
StudentId           object
dtype: object

In [24]:
sessions_DF.StudentClient.value_counts()

Website    158674
Android     77170
iOS         34798
Webapp      13299
Name: StudentClient, dtype: int64

In [20]:
sessions_DF.shape

(283941, 3)

In [9]:
sessions_DF.head(10)

SessionStartTime StudentClient  \
0  2017-11-18 15:47:33       Website   
1  2017-11-20 22:21:13       Website   
2  2017-11-20 22:35:31       Website   
3  2017-11-20 23:35:46       Website   
4  2017-11-23 21:24:00       Website   
5  2017-11-24 02:08:08       Website   
6  2017-11-13 13:12:49       Website   
7  2017-11-06 21:34:45       Website   
8  2017-11-20 15:26:09       Website   
9  2017-11-20 21:40:32       Website   

                                           StudentId  
0  0cade9bf00234e3789afc76250e4042bc1b70784ae7004...  
1  0cade9bf00234e3789afc76250e4042bc1b70784ae7004...  
2  0cade9bf00234e3789afc76250e4042bc1b70784ae7004...  
3  0cade9bf00234e3789afc76250e4042bc1b70784ae7004...  
4  0cade9bf00234e3789afc76250e4042bc1b70784ae7004...  
5  0cade9bf00234e3789afc76250e4042bc1b70784ae7004...  
6  8a501cab6c0a5a7e98a86f51a6264d81cdb562f4e99106...  
7  b8a39150d98d746853087a4107e9290976cb6c99d16a29...  
8  b8a39150d98d746853087a4107e9290976cb6c99d16a29...  
9  b8a39150d98d746853087a4107e9290976cb6c99d16a29...

Modificar SessionStartTime para timestamp e StudentClient para categórico.

In [34]:
subjects_DF.dtypes

Id       int64
Name    object
dtype: object

In [10]:
universities_DF.dtypes

Id       int64
Name    object
dtype: object

In [21]:
universities_DF.shape

(2076, 2)

In [11]:
universities_DF.head(10)

Id         Name
0  664704         UERJ
1  661625      PUC-RIO
2  664768       UNIRIO
3  663106         UFSJ
4  664138         UFSC
5  664742         UFRJ
6  663609       PUC-PR
7  663054         UFOP
8  663065          UFV
9  664565  ESTÁCIO EAD

In [12]:
subscriptions_DF.dtypes

PaymentDate    object
PlanType       object
StudentId      object
dtype: object

In [13]:
subscriptions_DF.head(10)

PaymentDate PlanType  \
0  2017-11-14 19:52:36.756325   Mensal   
1  2017-11-08 11:52:09.830538   Mensal   
2  2017-11-05 21:27:47.793878   Mensal   
3  2017-11-15 14:36:42.348867   Mensal   
4  2017-11-12 22:19:54.575387   Mensal   
5  2017-11-22 01:03:09.059587   Mensal   
6  2017-11-24 19:03:49.902349   Mensal   
7  2017-11-11 21:01:59.149344   Mensal   
8  2017-11-12 16:41:19.995949   Mensal   
9  2017-11-21 11:52:55.082745   Mensal   

                                           StudentId  
0  29037b0a52c5b576d32ac5842fd5b7e5991c5bef93150f...  
1  b2bace77d15c3dfaf57bf9916f0e9924d12f89bdc0624c...  
2  f423d6fe2f8964db634c7806e03b0f6567e992e994f06f...  
3  55ccbe518d2edbbd527eddca1431c7f7f3a4676bdbee38...  
4  b1b0f63fe3e4820cb082531af7c8911cd380e5ec8fe822...  
5  ed46832f6b716fb2e2f6f11b43965f2e0a3f2f8f8994ce...  
6  69b7bee32821cf76b9ffd5deffca1a01c95499cea24a45...  
7  6553923125fe6364e16aacf5656426809bb134b005f182...  
8  3903a334d1af8ce838bc2b7f497f11c40437573c1ad30b...  
9  4487f81a4ea9b3c3cf58aa31ecc21c30d6fb186993af61...

In [25]:
subscriptions_DF.PlanType.nunique()

2

In [22]:
subscriptions_DF.shape

(799, 3)

# Tarefa A

Para ser uma base para B.I. acho que a melhor forma seria ter agregações relevantes sobre os usuários de determinada categorias, então eu segmentei os usuários pelas seguintes dimensões:
N - Coluna (Qtd de Valores Distintos)

1 - PlanType (2) (done) <- Tipo de plano que o estudante está cadastrado
3 - State (27) (done) <- Estado que o estudante está cadastrado
4 - Student Client (4) (done) <- Client de onde o usuário está rodando aquela sessão
6 - UniversityId (2k+) (done) <- Id da Universidade
7 - CourseId (471) <- Id do curso que o subject Id pertence
8 - SubjectId (done) <- Id do Subject
9 - dt (1 to 10000) <- data em que o job rodou

Isso gera um cubo relativamente grande mas útil para várias análises e segmentações de usuários.

Para cada uma das combinações as seguintes variáveis seriam calculadas: 
* Quantidade de Sessões iniciadas (done)
* Quantidade de cadastros/pagamentos feitas (done)
* Quantidade de inscrições feitas (done)
* Quantidade de seguidores daquele Subject 

# Pipeline dos dados

O pipeline dos dados eu provavelmente faria em Spark em forma de batch dado que a quantidade de dados não é grande e não acho que a haja a necessidade de ser real-time (ainda).

Assumindo que o pipeline de dados é real-time eu alimentaria tópicos do Apache Kafka (imagine tópicos como sendo uma middleware do tipo fila de mensagens) onde diariamente estes dados seriam lançados para um armazenamento mais persistente ou robusto (Banco Relacional ou AWS S3, por exemplo) em um formato que fosse bem comprimido como Parquet ou Avro, ambos binários ao invés de um CSV ou JSON, por exemplo  e particionados por DT (data) dessa forma tornando fácil fazer jobs para cálculos retroativos ou em casos de retry (onde o job falha por algum motivo externo).

Desta forma, criaria jobs (pode ser Spark, pandas ou Dask, depende do volume de dados de fato, às vezes utilizar Spark pode ser um overkill para situação e atrasar a execução, gastando mais R$ com recursos) onde ele criaria o cubo, a forma como este job seria escrito provavelmente seria igual a seção "Data Preparation" abaixo dado que seria um grande conjunto de transformações e joins, como eu duvido que esse students_DF seja o dado que venha direto da aplicação e sim um compilado já salvo em um banco eu começaria com este ponto de verdade.

# Data Preparation

Primeiro utilizei os dados que espero que sejam os mais estáticos

## Aggregations

In [82]:
sessions_DF.columns

Index(['StudentClient', 'StudentId'], dtype='object')

In [75]:
# No caso dos estudantes serem seguidos, acho que seria melhor para o cubo que os dados fossem processados antes
# e depois juntados por SUBJECT_ID
student_follow_subject_DF["FollowDate"] = pandas.to_datetime(student_follow_subject_DF["FollowDate"])

KeyError: 'FollowDate'

In [50]:
sessions_DF["SessionStartTime"] = pandas.to_datetime(sessions_DF["SessionStartTime"])

In [52]:
sessions_DF = sessions_DF.set_index("SessionStartTime")

In [55]:
sessions_daily_counter_DF = sessions_DF.groupby("StudentClient").resample('D').agg({'StudentId': 'count'})

In [62]:
sessions_daily_counter_DF.reset_index(inplace=True)

In [56]:
sessions_daily_counter_DF.head(10)

StudentId
StudentClient SessionStartTime           
Android       2017-11-01             1609
              2017-11-02             1148
              2017-11-03             1432
              2017-11-04             1417
              2017-11-05             1634
              2017-11-06             2077
              2017-11-07             2351
              2017-11-08             2223
              2017-11-09             2245
              2017-11-10             2281

In [76]:
sessions_daily_counter_DF.head(10)

StudentClient SessionStartTime  StudentId
0       Android       2017-11-01       1609
1       Android       2017-11-02       1148
2       Android       2017-11-03       1432
3       Android       2017-11-04       1417
4       Android       2017-11-05       1634
5       Android       2017-11-06       2077
6       Android       2017-11-07       2351
7       Android       2017-11-08       2223
8       Android       2017-11-09       2245
9       Android       2017-11-10       2281

In [78]:
sessions_daily_counter_DF.columns

Index(['StudentClient', 'SessionStartTime', 'StudentId'], dtype='object')

In [34]:
student_follow_subject_DF = student_follow_subject_DF.set_index("FollowDate")

KeyError: 'FollowDate'

In [41]:
subject_follow_agg = student_follow_subject_DF.groupby("SubjectId").resample('D').agg({'StudentId': 'count'})

In [70]:
subject_follow_agg.head(10)

StudentId
SubjectId FollowDate           
669382    2017-11-15          1
669383    2015-09-05          1
          2015-09-06          0
          2015-09-07          0
          2015-09-08          1
          2015-09-09          0
          2015-09-10          0
          2015-09-11          0
          2015-09-12          0
          2015-09-13          0

In [71]:
subject_follow_agg.reset_index(inplace=True)

In [72]:
subject_follow_agg.head(10)

SubjectId FollowDate  StudentId
0     669382 2017-11-15          1
1     669383 2015-09-05          1
2     669383 2015-09-06          0
3     669383 2015-09-07          0
4     669383 2015-09-08          1
5     669383 2015-09-09          0
6     669383 2015-09-10          0
7     669383 2015-09-11          0
8     669383 2015-09-12          0
9     669383 2015-09-13          0

## Joins

In [47]:
subject_follow_agg["StudentId"].head(10)

SubjectId  FollowDate
669382     2017-11-15    1
669383     2015-09-05    1
           2015-09-06    0
           2015-09-07    0
           2015-09-08    1
           2015-09-09    0
           2015-09-10    0
           2015-09-11    0
           2015-09-12    0
           2015-09-13    0
Name: StudentId, dtype: int64

In [ ]:
subject_stats = pandas.merge(subjects_DF)

In [81]:
temp_DF = pandas.merge(students_DF, subscriptions_DF, how="left", left_on = ["Id"], right_on = ["StudentId"]).drop("StudentId", axis=1)
temp_DF.columns
#temp_DF = pandas.merge(temp_DF, sessions_daily_counter_DF.drop("SessionStartTime", axis=1), how="left", left_on=["SubjectId"], right_on=["SubjectId"])
#temp_DF = pandas.merge(temp_DF, student_follow_subject_DF, how="left", left_on = ["Id"], right_on = ["StudentId"]).drop("StudentId", axis=1)

Index(['City', 'CourseId', 'Id', 'RegisteredDate', 'SignupSource', 'State',
       'StudentClient', 'UniversityId', 'PaymentDate', 'PlanType'],
      dtype='object')

In [20]:
temp_DF.dtypes

City               object
CourseId            int64
Id                 object
RegisteredDate     object
SignupSource       object
State              object
StudentClient      object
UniversityId        int64
PaymentDate        object
PlanType           object
FollowDate         object
SubjectId         float64
dtype: object

Por algum motivo o tipo de SubjecTId é modificado mas é só colocar para ler como int64 novamente.